File: oc/UIPickerViewDemo/UIPickerViewDemo/ProvinceViewController.m
//
//  ProvinceViewController.m
//  UIPickerViewDemo
//
//  Created by dfang on 2018-1-6.
//  Copyright © 2018年 east. All rights reserved.
//

#import "ProvinceViewController.h"

@interface ProvinceViewController ()

@property (nonatomic, strong) UIPickerView *pickerView;
@property (nonatomic, strong) NSArray *provinces;
@property (nonatomic, assign) NSInteger provinceSelected;

@end

@implementation ProvinceViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.pickerView.delegate = self;
    self.pickerView.dataSource = self;
    [self.view addSubview:self.pickerView];
    
    _provinceSelected = 0;
    for (int i = 0; i < self.provinces.count; i++) {
        [self pickerView:self.pickerView didSelectRow:0 inComponent:i];
    }
}

- (UIPickerView *)pickerView {
    if (_pickerView == nil) {
        _pickerView = [[UIPickerView alloc] initWithFrame:self.view.frame];
    }
    return _pickerView;
}

- (NSArray *)provinces {
    if (_provinces == nil) {
        _provinces = [NSArray arrayWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"provinces" ofType:@"plist"]];
    }
    
    return _provinces;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 2;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    if (component == 0) {
        return self.provinces.count;
    } else {
        NSDictionary *item = self.provinces[self.provinceSelected];
        NSArray *cities = [item valueForKey:@"cities"];
        return cities.count;
    }
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    if (component == 0) {
        NSDictionary *item = self.provinces[row];
        return [item valueForKey:@"name"];
    } else {
        NSDictionary *item = self.provinces[self.provinceSelected];
        NSArray *cities = [item valueForKey:@"cities"];
        return cities[row];
    }
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    if (component == 0) {
        self.provinceSelected = row;
        
        [self.pickerView reloadComponent:1];
        [self.pickerView selectRow:0 inComponent:1 animated:true];
    }
}

@end

File: swift/UIPickerViewDemo/UIPickerViewDemo/ProvinceViewController.swift
//
//  ProvinceViewController.swift
//  UIPickerViewDemo
//
//  Created by dfang on 2018-1-5.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class ProvinceViewController: UIViewController, UIPickerViewDataSource, UIPickerViewDelegate {
    
    let pickerView = UIPickerView.init()
    let provinceList = Province.provinceList()
    var provinceSelected = 0
    
    override func viewDidLoad() {
        pickerView.frame = self.view.frame
        pickerView.dataSource = self
        pickerView.delegate = self
        self.view.addSubview(pickerView)
        
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 2
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        switch component {
        case 0:
            return provinceList.count
        case 1:
            let province = provinceList[provinceSelected] as! Province
            return province.cities.count
        default:
            return 0
        }
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        if (component == 0) {
            let province = provinceList[row] as! Province
            return province.name
        } else {
            let province = provinceList[provinceSelected] as! Province
            return province.cities.object(at: row) as? String
        }
    }
    
    /*
    func pickerView(_ pickerView: UIPickerView, attributedTitleForRow row: Int, forComponent component: Int) -> NSAttributedString? {
        if (component == 0) {
            let province = provinceList[row] as! Province
            let attributeStr = NSAttributedString.init(string: province.name, attributes: [NSAttributedStringKey.foregroundColor:UIColor.red])
            return attributeStr
        } else {
            let province = provinceList[provinceSelected] as! Province
            let str = province.cities.object(at: row) as! String
            return NSAttributedString.init(string: str, attributes: [NSAttributedStringKey.foregroundColor:UIColor.blue])
        }
    }
    */

    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        if (component == 0) {
            provinceSelected = row
            pickerView.reloadComponent(1)
            pickerView.selectRow(0, inComponent: 1, animated: true)
        }
    }
}
